storagenode/piecestore/endpoint.go

```diff
@@ -314,7 +334,46 @@ func (endpoint *Endpoint) Upload(stream pb.DRPCPiecestore_UploadStream) (err err
 	largestOrder := pb.Order{}
 	defer commitOrderToStore(ctx, &largestOrder)
 
+	// Measure upload rate and average upload rate
+	uploadSize := int64(0)
+	previousSize := int64(0)
+
+	// sumUploadRate sums all previously sampled upload rate
+	sumUploadRate := float64(0)
+	// averageUploadRate= sum of sampled upload rate / total of samples
+	averageUploadRate := float64(0)
+	// number as of data chunks that have been received so far
+	// first chunk is counted as 0
+	chunkCount := float64(-1)
+
 	for {
+		// Increment counts
+		chunkCount++
+		// Measure upload rate (bytes per second)
+		currentTime := time.Now().UTC()
+		dt := currentTime.Sub(startTime)
+
+		if pieceWriter != nil {
+			// delta uploaded
+			uploadSize = pieceWriter.Size() - previousSize
+			previousSize = pieceWriter.Size()
+		}
+
+		if dt.Seconds() > 0 {
+			// currentUploadRate mesures the rate of upload for each chunk of data.
+			currentUploadRate := float64(uploadSize) / dt.Seconds()
+			sumUploadRate += currentUploadRate
+
+			// Average upload rate is counted as: total upload rate samples per a period of time
+			// divides by total number of chunks
+			averageUploadRate = sumUploadRate / chunkCount
+		}
+
+		// Verify if the upload is supiciously slow
+		if endpoint.flagSuspiciousSlowUpload(averageUploadRate, int64(liveRequests)) {
```

Comment:
We should probably wait for a few seconds after a transfer starts before activating this check. The `averageUploadRate` (or `currentUploadRate`) value will tend to fluctuate wildly over the first few seconds, as data does not arrive at a perfectly steady rate. Perhaps, if `dt` is less than, say, 10 seconds then we skip all these calculations and checks entirely?

Reply:
Fixed